File: src/json/matias_ergo_pro.json.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require_relative '../lib/karabiner.rb'
require_relative '../lib/key_codes.rb'
require_relative '../lib/variable.rb'

def main
  puts JSON.pretty_generate(
    'title' => Device::NAME,
    'maintainers' => ['mingaldrichgan'],
    'rules' => [Rules::NavKeys, Rules::RightControl].map(&:rule)
  )
end

module Device
  NAME = 'Matias Ergo Pro Keyboard'

  CONDITION = {
    'description' => NAME,
    'type' => 'device_if',
    'identifiers' => [{ 'vendor_id' => 1452, 'product_id' => 591 }],
  }.freeze

  TOP_ROW = (KeyCodes.f(1..12) + %w[escape delete_forward]).freeze
  TYPING_KEYS = (KeyCodes::ALPHANUMERIC_ROWS + ['spacebar']).freeze
end

module Rules
  module NavKeys
    TO_KEY = 'right_option'
    VAR = Variable.new("use_nav_keys_as_#{TO_KEY}").freeze

    NAV_KEYS = %w[home end page_up page_down].freeze
    NAV_MODIFIERS = %w[right_command right_shift].freeze

    def self.rule
      {
        'description' => "#{Device::NAME}: Change navigation keys to #{TO_KEY} if pressed with another key",
        'manipulators' => VAR.remap_to_modifiers(
          NAV_KEYS,
          to_modifiers: [TO_KEY],
          modifiable: Device::TOP_ROW + Device::TYPING_KEYS + NAV_KEYS + NAV_MODIFIERS + KeyCodes::ARROWS
        ),
      }
    end
  end # module NavKeys

  module RightControl
    FROM_KEY = 'right_control'
    TO_KEY = 'b'
    VAR = Variable.new("use_#{FROM_KEY}_as_#{TO_KEY}").freeze

    TYPING_MODIFIERS = Karabiner.from_modifiers(nil, %w[caps_lock shift]).freeze

    def self.rule
      {
        'description' => "#{Device::NAME}: Change #{FROM_KEY} to #{TO_KEY} if pressed alone, or while typing",
        'manipulators' => (
          Device::TYPING_KEYS.map do |key|
            {
              'description' => "Set variable if #{key} is pressed",
              'type' => 'basic',
              'from' => { 'key_code' => key, 'modifiers' => TYPING_MODIFIERS },
              'to' => [VAR.set, { 'key_code' => key }],
              'to_delayed_action' => { 'to_if_invoked' => [VAR.unset] },
              'conditions' => [Device::CONDITION, NavKeys::VAR.unless_set],
            }
          end + [
            {
              'description' => "Change #{FROM_KEY} to #{TO_KEY} if variable is set",
              'type' => 'basic',
              'from' => { 'key_code' => FROM_KEY, 'modifiers' => Karabiner.from_modifiers },
              'to' => [VAR.set, { 'key_code' => TO_KEY }],
              'to_delayed_action' => { 'to_if_invoked' => [VAR.unset] },
              'conditions' => [Device::CONDITION, VAR.if_set],
            },
            {
              'description' => "Change #{FROM_KEY} to #{TO_KEY} if pressed alone and not held down",
              'type' => 'basic',
              'from' => { 'key_code' => FROM_KEY, 'modifiers' => Karabiner.from_modifiers },
              'to' => [{ 'key_code' => FROM_KEY, 'lazy' => true }],
              'to_if_alone' => [VAR.set, { 'key_code' => TO_KEY }],
              'to_if_held_down' => [{ 'key_code' => FROM_KEY }],
              'to_delayed_action' => { 'to_if_invoked' => [VAR.unset] },
              'conditions' => [Device::CONDITION, VAR.unless_set],
              'parameters' => {
                # Set these parameters to the same value.
                'basic.to_if_alone_timeout_milliseconds' => 500, # Default value is 1000.
                'basic.to_if_held_down_threshold_milliseconds' => 500, # Default value.
              },
            },
          ]
        ),
      }
    end
  end # module RightControl
end # module Rules

main

File: src/json/matias_ergo_pro.json.js
// JavaScript should be written in ECMAScript 5.1.

const karabiner = require('../lib/karabiner')

function main() {
  console.log(
    JSON.stringify(
      {
        title: device.name,
        maintainers: ['mingaldrichgan'],
        rules: [navKeys(), rightControl()],
      },
      null,
      '  '
    )
  )
}

const device = {
  name: 'Matias Ergo Pro Keyboard',

  condition: {
    description: 'Matias Ergo Pro Keyboard',
    type: 'device_if',
    identifiers: [{ vendor_id: 1452, product_id: 591 }],
  },

  navToKey: 'right_option',
  navVariable: function () {
    return 'matias_ergo_pro.use_nav_keys_as_' + device.navToKey
  },

  topRow: ['f1', 'f2', 'f3', 'f4', 'f5', 'f6', 'f7', 'f8', 'f9', 'f10', 'f11', 'f12', 'escape', 'delete_forward'],

  typingKeys: [].concat(
    karabiner.numbers,
    karabiner.letters,
    ['grave_accent_and_tilde', 'hyphen', 'equal_sign', 'delete_or_backspace', 'tab'],
    ['open_bracket', 'close_bracket', 'backslash', 'semicolon', 'quote'],
    ['return_or_enter', 'comma', 'period', 'slash', 'spacebar']
  ),
}

function navKeys() {
  const toKey = device.navToKey
  const navKeys = ['home', 'end', 'page_up', 'page_down']
  const navModifiers = ['right_command', 'right_shift']
  const variable = device.navVariable()

  return {
    description: device.name + ': Change navigation keys to ' + toKey + ' if pressed with another key',
    manipulators: remapToModifiers({
      fromKeyCodes: navKeys,
      fromModifiers: { optional: ['any'] },
      toModifiers: [toKey],
      modifiable: [].concat(device.topRow, device.typingKeys, navKeys, navModifiers, karabiner.arrows),
      variable,
    }),
  }
}
function remapToModifiers(options) {
  const modifiers = [].concat(
    ['caps_lock', 'fn', 'command', 'control', 'option', 'shift'],
    ['left_command', 'left_control', 'left_option', 'left_shift', 'left_alt', 'left_gui'],
    ['right_command', 'right_control', 'right_option', 'right_shift', 'right_alt', 'right_gui']
  )

  return [].concat(
    combination(options.fromKeyCodes).map(function (pair) {
      return {
        description: 'Set variable if ' + pair.join(' + ') + ' are pressed simultaneously',
        type: 'basic',
        from: {
          modifiers: options.fromModifiers,
          simultaneous: pair.map(function (key) {
            return { key_code: key }
          }),
          simultaneous_options: {
            to_after_key_up: [{ set_variable: { name: options.variable, value: 0 } }],
          },
        },
        to: [{ set_variable: { name: options.variable, value: 1 } }],
        conditions: [{ type: 'variable_unless', name: options.variable, value: 1 }],
      }
    }),

    product(
      options.fromKeyCodes,
      options.modifiable.filter(function (m) {
        return options.fromKeyCodes.indexOf(m) === -1
      })
    ).map(function (pair) {
      return {
        description: 'Change ' + pair[0] + ' to ' + options.toModifiers.join(' + ') + ' if pressed with ' + pair[1],
        type: 'basic',
        from: {
          modifiers: options.fromModifiers,
          simultaneous: pair.map(function (key) {
            return { key_code: key }
          }),
          simultaneous_options: {
            to_after_key_up: [{ set_variable: { name: options.variable, value: 0 } }],
          },
        },
        to: [{ set_variable: { name: options.variable, value: 1 } }, { key_code: pair[1], modifiers: options.toModifiers }],
        conditions: [{ type: 'variable_unless', name: options.variable, value: 1 }],
      }
    }),

    options.modifiable
      .filter(function (m) {
        return modifiers.indexOf(m) === -1
      })
      .map(function (key) {
        return {
          description: 'Modify ' + key + ' with ' + options.toModifiers.join(' + ') + ' if variable is set',
          type: 'basic',
          from: { key_code: key, modifiers: options.fromModifiers },
          to: [{ key_code: key, modifiers: options.toModifiers }],
          conditions: [{ type: 'variable_if', name: options.variable, value: 1 }],
        }
      })
  )
}

function rightControl() {
  const fromKey = 'right_control'
  const toKey = 'b'
  const variable = 'matias_ergo_pro.use_' + fromKey + '_as_' + toKey
  const typingModifiers = { optional: ['caps_lock', 'shift'] }

  return {
    description: device.name + ': Change ' + fromKey + ' to ' + toKey + ' if pressed alone, or while typing',
    manipulators: [].concat(
      device.typingKeys.map(function (key) {
        return {
          description: 'Set variable if ' + key + ' is pressed',
          type: 'basic',
          from: { key_code: key, modifiers: typingModifiers },
          to: [
            {
              set_variable: { name: variable, value: 1 },
            },
            { key_code: key },
          ],
          to_delayed_action: {
            to_if_invoked: [{ set_variable: { name: variable, value: 0 } }],
          },
          conditions: [
            device.condition,
            {
              type: 'variable_unless',
              name: device.navVariable(),
              value: 1,
            },
          ],
        }
      }),
      [
        {
          description: 'Change ' + fromKey + ' to ' + toKey + ' if variable is set',
          type: 'basic',
          from: { key_code: fromKey, modifiers: { optional: ['any'] } },
          to: [
            {
              set_variable: { name: variable, value: 1 },
            },
            { key_code: toKey },
          ],
          to_delayed_action: { to_if_invoked: [{ set_variable: { name: variable, value: 0 } }] },
          conditions: [
            device.condition,
            {
              type: 'variable_if',
              name: variable,
              value: 1,
            },
          ],
        },
        {
          description: 'Change ' + fromKey + ' to ' + toKey + ' if pressed alone and not held down',
          type: 'basic',
          from: { key_code: fromKey, modifiers: { optional: ['any'] } },
          to: [{ key_code: fromKey, lazy: true }],
          to_if_alone: [
            {
              set_variable: { name: variable, value: 1 },
            },
            { key_code: toKey },
          ],
          to_if_held_down: [{ key_code: fromKey }],
          to_delayed_action: { to_if_invoked: [{ set_variable: { name: variable, value: 0 } }] },
          conditions: [
            device.condition,
            {
              type: 'variable_unless',
              name: variable,
              value: 1,
            },
          ],
          parameters: {
            // Set these parameters to the same value.
            'basic.to_if_alone_timeout_milliseconds': 500, // Default value is 1000.
            'basic.to_if_held_down_threshold_milliseconds': 500, // Default value.
          },
        },
      ]
    ),
  }
}

function combination(array) {
  return array.reduce(function (acc, a, i) {
    return acc.concat(
      array.slice(i + 1).map(function (b) {
        return [a, b]
      })
    )
  }, [])
}

function product(a, b) {
  return a.reduce(function (acc, x) {
    return acc.concat(
      b.map(function (y) {
        return [x, y]
      })
    )
  }, [])
}

main()
